quinn/src/lib.rs

```diff
@@ -879,16 +873,34 @@ impl Future for Driver {
             let mut fired = false;
             loop {
                 match endpoint.timers.poll() {
-                    Ok(Async::Ready(Some(Some((conn, timer))))) => {
+                    Ok(Async::Ready(Some(expired))) => {
+                        let (conn, timer) = expired.into_inner();
                         trace!(endpoint.log, "timeout"; "timer" => ?timer);
                         endpoint.inner.timeout(now, conn, timer);
                         fired = true;
                     }
-                    Ok(Async::Ready(Some(None))) => {}
+                    // Ready(None) is returned if there's not active timer, and NotReady is
+                    // returned if there are active timers but none expired yet.
                     Ok(Async::Ready(None)) | Ok(Async::NotReady) => {
                         break;
                     }
-                    Err(()) => unreachable!(),
+                    Err(timer_err) => {
+                        if timer_err.is_shutdown() {
+                            // This miht only happen is there's a bug in some low level tokio
+                            // components.
+                            panic!("{}", timer_err);
+                        }
+                        if timer_err.is_at_capacity() {
+                            // This occurs if there are too many Delay instances. But the delay
+                            // queue uses a single timer so there's not much we can do here in
+                            // terms of backpressure.
+                            panic!("{}", timer_err);
+                        }
+                        // There are only two types of error in tokio-timer, but when we upgrade,
+                        // there might be more, so let's panic here, in case we forget to update
+                        // this code,
```

Comment:
Misleading comment, since errors can be introduced without a breaking change and therefore without us having upgraded anything.